engine/classes/Elgg/Http/ResponseFactory.php

```diff
@@ -0,0 +1,460 @@
+<?php
+
+namespace Elgg\Http;
+
+use Elgg\Ajax\Service as AjaxService;
+use Elgg\PluginHooksService;
+use ElggEntity;
+use InvalidArgumentException;
+use InvalidParameterException;
+use SecurityException;
+use Symfony\Component\HttpFoundation\RedirectResponse as SymfonyRedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+
+/**
+ * WARNING: API IN FLUX. DO NOT USE DIRECTLY.
+ *
+ * @since 2.2.0
+ * @access private
+ */
+class ResponseFactory {
+
+	/**
+	 * @var Request
+	 */
+	private $request;
+
+	/**
+	 * @var AjaxService
+	 */
+	private $ajax;
+
+	/**
+	 * @var PluginHooksService
+	 */
+	private $hooks;
+
+	/**
+	 * @var Response|bool
+	 */
+	private $response_sent = false;
+
+	/**
+	 * @var ResponseHeaderBag
+	 */
+	private $headers;
+
+	/**
+	 * Constructor
+	 * 
+	 * @param Request            $request HTTP request
+	 * @param PluginHooksService $hooks   Plugin hooks service
+	 * @param AjaxService        $ajax    AJAX service
+	 */
+	public function __construct(Request $request, PluginHooksService $hooks, AjaxService $ajax) {
+		$this->request = $request;
+		$this->hooks = $hooks;
+		$this->ajax = $ajax;
+		$this->headers = new ResponseHeaderBag();
+	}
+
+	/**
+	 * Sets headers to apply to all responses being sent
+	 *
+	 * @param string $name    Header name
+	 * @param string $value   Header value
+	 * @param bool   $replace Replace existing headers
+	 * @return void
+	 */
+	public function setHeader($name, $value, $replace = true) {
+		$this->headers->set($name, $value, $replace);
+	}
+
+	/**
+	 * Get headers set to apply to all responses
+	 * @return ResponseHeaderBag
+	 */
+	public function getHeaders() {
```

Comment:
It picks up the headers set by Sessions, which is good, but the response should *overwrite* the existing headers.